OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs

```diff
@@ -21,7 +21,19 @@ namespace OpenRA.Mods.Common.Traits
 	[Desc("Attach this to a unit to enable dynamic upgrades by warheads, experience, crates, support powers, etc.")]
 	public class UpgradeManagerInfo : ITraitInfo, Requires<IUpgradableInfo>
 	{
-		public object Create(ActorInitializer init) { return new UpgradeManager(init); }
+		[UpgradeGrantedReference]
+		[Desc("Upgrades that can be granted/revoked from Lua scripts.")]
+		public readonly HashSet<string> Scriptable = new HashSet<string>();
+
+		[UpgradeUsedReference]
+		[Desc("Upgrades not to check whether used.")]
+		public readonly HashSet<string> IgnoredAsUnused = new HashSet<string>();
```

Comment:
I don't know whether to call this a hack or a straight-up bug, but it is bad. (I actually still haven't heard a reason to have it)